subsys/shell/shell_uart.c

```diff
@@ -27,62 +27,68 @@ SHELL_DEFINE(shell_uart, CONFIG_SHELL_PROMPT_UART, &shell_transport_uart,
 	     SHELL_FLAG_OLF_CRLF);
 
 #ifdef CONFIG_SHELL_BACKEND_SERIAL_INTERRUPT_DRIVEN
+static inline bool smp_rx_byte(const struct shell_uart *sh_uart, uint8_t byte)
+{
+#ifdef CONFIG_MCUMGR_SMP_SHELL
+	return smp_shell_rx_byte(&sh_uart->ctrl_blk->smp, byte);
+#else
+	return false;
+#endif
+}
+
 static void uart_rx_handle(struct device *dev,
 			   const struct shell_uart *sh_uart)
 {
 	uint8_t *data;
+	uint8_t dummy;
 	uint32_t len;
+	uint32_t to_read;
 	uint32_t rd_len;
+	uint32_t rd_idx;
+	uint32_t uart_idx;
 	bool new_data = false;
 
 	do {
 		len = ring_buf_put_claim(sh_uart->rx_ringbuf, &data,
 					 sh_uart->rx_ringbuf->size);
-
 		if (len > 0) {
-			rd_len = uart_fifo_read(dev, data, len);
-#ifdef CONFIG_MCUMGR_SMP_SHELL
-			/* Divert bytes from shell handling if it is
-			 * part of an mcumgr frame.
-			 */
-			size_t i;
-
-			for (i = 0; i < rd_len; i++) {
-				if (!smp_shell_rx_byte(&sh_uart->ctrl_blk->smp,
-						       data[i])) {
-					break;
-				}
-			}
+			to_read = len;
+		} else {
+			LOG_WRN("RX ring buffer full.");
 
-			rd_len -= i;
-			new_data = true;
-			if (rd_len) {
-				for (uint32_t j = 0; j < rd_len; j++) {
-					data[j] = data[i + j];
-				}
-			}
-#else
-			if (rd_len > 0) {
-				new_data = true;
+			to_read = 1;
+			data = &dummy;
+		}
+
+		rd_len = uart_fifo_read(dev, data, to_read);
+		rd_idx = 0;
+		uart_idx = 0;
+
+		while (rd_idx < rd_len) {
+			bool consumed = smp_rx_byte(sh_uart, data[rd_idx]);
+
+			if (!consumed) {
+				data[uart_idx++] = data[rd_idx];
 			}
-#endif /* CONFIG_MCUMGR_SMP_SHELL */
+
+			rd_idx++;
+		}
```

Comment:
This can benefit from https://github.com/zephyrproject-rtos/zephyr/pull/27657, I do not think there is need to pass each byte one by one to be processed. We may assume that SMP bytes will fallow in bulk so invoking function for each byte instead of giving it part of buffer to process seems wasteful.

Reply:
Yes, I agree.

Reply:
Additionally, I do not think that the `while` loop will be optimized out when `smp_rx_byte` would be compiled to return `false`, as `rd_idx++` has side effects. Compiler may figure out to do `rd_idx += rd_len - rd_idx - 1` instead of loop but I doubt it will. Anyway, it will leave some code.